### ISSUE TITLE
Make the default source build the most optimized

### ISSUE THREAD

**Issue by certik** [symengine/symengine#1579]
This is related to the recent removal of `-march=native` (e.g., in #1577).

My design idea early on was to always enable the most optimized build when building from source. The reason for that is that from my experience, when people download the source of SymEngine, even if I told them explicitly in the email how to enable optimizations, they still don't do that. Right now we have it documented in the README, but I bet most people will not bother reading it and will just install using the usual cmake approach.

For this reason, I really think the default cmake build should be the most optimized (including `-march=native` and such).

For distributions like Conda, Debian, etc., we should have a dedicated section in the README, how to configure our cmake to build for distributions.

There is an issue that I don't know if CMake allows to remove `-march=native` when passed some command line options, but I think it does, I think that was the motivation to use `UserOverride.cmake`, which just provides the defaults, but if the user specifies `CMAKE_CXX_FLAGS_RELEASE`, it will override the defaults (i.e., not use `-march=native`).  so I think it's technically solvable.

Essentially we have to decide whether our default should be for distributions, or for end users. I feel end users should be prioritized. We can help distributions build it correctly, but it should just work for end users. Also typically distributions actually consult the README to see how to build it correctly.

**Comment by bjodah** [comment]
What about pointing them to a one-liner along the lines of:

```bash
$ curl -Ls https://github.com/symengine/symengine/download_compile_and_install_symengine_interactive.sh | bash
```

and then have the script check dependencies and ask the user for options interactively? (and a user pressing enter throughout would get some optimized but perhaps not portable defaults?)

The downside is that there could be a lot of work providing support and fixing a slew of corner cases for different users environments...

**Comment by certik** [comment]
@bjodah that's one option, but as you said, more work for us.

Although we are already dedicated to support any user's setup with our CMake, so far I think we always managed to fix it, so I think that's not a problem to ensure that our CMake itself works.

Supporting three scripts on Linux, macOS and Windows is another issue. I would say let's just use CMake, instead of a script.


The main question is: by default, do we want things to just work for end users installing from source, or for distributors like Conda?

In my mind, it's for sure end users. Distributors can enable extra options easily in their package scripts. While end users just want to run CMake without options and things to just work.

**Comment by bjodah** [comment]
Sure, but removing defaults is harder than adding stuff in e.g. CXXFLAGS, for example I have this script for my personal usage (note my call to `sed` on the highlighted line):
https://github.com/bjodah/bjodahimg18dev/blob/master/environment/scripts/get_symengine.sh#L30

**Comment by certik** [comment]
Yes, it's a bug if you have to run sed on symengine sources. As I mentioned above, you remove defaults by simply providing options on the command line of cmake. Does it not work?

**Comment by bjodah** [comment]
Umm, not sure at this point. But without the highlighted line (at the time I last edited the script) I got binaries compiled with `-march=native` (as we've already discussed, and I see Isuru is working on the final PR on the matter). Either way, there's a trade off for you to make and I'll adapt to whatever you decide :)

**Comment by certik** [comment]
Ok, I tested it and it works. Here is how: in the current master (4f7cd0e74901e4e6cabcfab2cdcbd72c85d455ef) if you do:
```
$ cmake .
...
CMAKE_CXX_FLAGS_RELEASE: -Wall -Wextra -Wno-unused-parameter -fno-common -O3 -march=native -funroll-loops -std=c++11 -fPIC -Wno-unknown-pragmas
...
```
You get the `-march=native` in there. To remove it (and also to use `O2` instead of `O3`), just do:
```
$ cmake -DCMAKE_CXX_FLAGS_RELEASE="-O2" .
...
CMAKE_CXX_FLAGS_RELEASE: -O2 -std=c++11 -fPIC -Wno-unknown-pragmas
```
To remove all flags completely, just do:
```
$ cmake -DCMAKE_CXX_FLAGS_RELEASE= .
...
CMAKE_CXX_FLAGS_RELEASE:  -std=c++11 -fPIC -Wno-unknown-pragmas
```
The only flags that remain there are those that are actually require to compile the code no matter what.

**Comment by bjodah** [comment]
Oh, great. Thanks! I forgot about that flag, (cmake usage really is an acquired taste, but it's best in class when it comes to supporting platforms/compilers for sure).

**Comment by isuruf** [comment]
`-march=native` doesn't make a visible change in benchmarks for symengine as mentioned in https://github.com/symengine/symengine/issues/1542. Therefore I don't see the point in enabling it.
